oracle-plugin/src/main/java/io/cdap/plugin/oracle/OracleDBRecord.java

```diff
@@ -50,31 +61,208 @@ protected SchemaReader getSchemaReader() {
+   * @return instance of 'oracle.sql.TIMESTAMPTZ' which corresponds to the specified timestamp with time zone string.
+   */
+  private Object createOracleTimestampWithTimeZone(Connection connection, String timestampString) {
+    try {
+      ClassLoader classLoader = connection.getClass().getClassLoader();
+      Class<?> timestampTZClass = classLoader.loadClass("oracle.sql.TIMESTAMPTZ");
+      return timestampTZClass.getConstructor(Connection.class, String.class).newInstance(connection, timestampString);
+    } catch (ClassNotFoundException e) {
+      throw new InvalidStageException("Unable to load 'oracle.sql.TIMESTAMPTZ'.", e);
+    } catch (InstantiationException | InvocationTargetException | NoSuchMethodException | IllegalAccessException e) {
+      throw new InvalidStageException("Unable to instantiate 'oracle.sql.TIMESTAMPTZ'.", e);
+    }
+  }
 
+  /**
+   * Creates an instance of 'oracle.sql.BFILE'. Note, that we can not create an operating system file that a 'BFILE'
+   * would refer to. <a href="https://docs.oracle.com/cd/B19306_01/java.102/b14355/oralob.htm#BABJJEIC">Those
+   * are created only externally.</a>
+   * @param connection sql connection.
+   * @param bytes BFILE locator's bytes.
+   * @return instance of 'oracle.sql.BFILE'.
+   */
+  private Object createOracleBfile(Connection connection, byte[] bytes) {
+    try {
+      ClassLoader classLoader = connection.getClass().getClassLoader();
+      Class<?> oracleConnectionClass = classLoader.loadClass("oracle.jdbc.OracleConnection");
+      Class<?> bfileClass = classLoader.loadClass("oracle.sql.BFILE");
+      return bfileClass.getConstructor(oracleConnectionClass, byte[].class).newInstance(connection, bytes);
+    } catch (ClassNotFoundException e) {
+      throw new InvalidStageException("Unable to load Oracle JDBC connector.", e);
+    } catch (InstantiationException | InvocationTargetException | NoSuchMethodException | IllegalAccessException e) {
+      throw new InvalidStageException("Unable to instantiate 'oracle.sql.BFILE'.", e);
+    }
+  }
+
+  /**
+   * Retrieves bytes representation of 'oracle.sql.BFILE' via 'oracle.sql.BFILE#getBytes'. Note, that this method
+   * retrieves only locator (link) to an external binary file (file stored outside of the database) and not content of
+   * the file. <a href="https://docs.oracle.com/cd/B19306_01/java.102/b14355/oralob.htm#BABJJEIC">Files are created
+   * only externally.</a>
+   * @param bfile instance of 'oracle.sql.BFILE'.
+   * @return bytes representation of 'oracle.sql.BFILE' obtained via 'oracle.sql.BFILE#getBytes'.
+   */
+  private byte[] getBfileBytes(ResultSet resultSet, Object bfile) {
+    if (bfile == null) {
+      return null;
+    }
+    try {
+      ClassLoader classLoader = resultSet.getClass().getClassLoader();
+      Class<?> bfileClass = classLoader.loadClass("oracle.sql.BFILE");
+      return (byte[]) bfileClass.getMethod("getBytes").invoke(bfile);
```

Comment:
> You said this returns a byte representation of the locator? I just don't see how that would be useful to anybody, why not return the contents of the file?

Maybe it will be useful when Source and Sink use Oracle tables of the same database. In other cases, the user will be forced to move external files manually. I was trying to represent only locators because Oracle's CLI does so.

BFILE mapping changed to retrieve actual file contents.